### PR TITLE
Adding VIB regularization in CQL

### PR DESCRIPTION
**Pull request by abhiksingla**
## Why are these changes needed?
The variational Information Bottleneck technique is another potential regularization method to resolve the unlearning issue in CQL.

- The regularize is configured by passing a dictionary in the trainer config using key "vib_regularizer"
- Setting "enable_regualarizer" in the dictionary activates the regularizer.
- Field "beta" sets the weight of information loss vs the cql loss.

Note: Files vib.py and vib_dev.py are for prototyping and will be removed during the cleanup.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/sac/sac_tf_model.py [hunk 1]
```diff
@@ -73,6 +79,7 @@ def __init__(self,
         self.model_out = tf.keras.layers.Input(
             shape=(self.num_outputs, ), name="model_out")
         self.action_model = self.build_action_model()
+        keras.utils.plot_model(self.action_model, "action_model.png", show_shapes=True)
```
Comment (Edilmo) [inline rllib/agents/sac/sac_tf_model.py]: let's remove this

File: rllib/agents/sac/sac_tf_model.py [hunk 2]
```diff
@@ -83,6 +90,7 @@ def __init__(self,
                 shape=(self.action_dim, ), name="actions")
 
         self.q_net = self.build_q_net("q", self.model_out, self.actions_input)
+        keras.utils.plot_model(self.q_net, "q_net.png", show_shapes=True)
```
Comment (Edilmo) [inline rllib/agents/sac/sac_tf_model.py]: let's remove this

File: rllib/agents/cql/cql_sac.py [hunk 3]
```diff
@@ -39,6 +39,16 @@
         # Replay Buffer should be size of offline dataset for fastest
         # training
         "buffer_size": 1000000,
+        # Upper bound for alpha value during the lagrangian constraint
+        "alpha_upper_bound": 1.0,
+        # Lower bound for alpha value during the lagrangian constraint
+        "alpha_lower_bound": 0.0,
+        # VIB regularization parameters
+        "vib_regularizer": {
```
Comment (Edilmo) [inline rllib/agents/cql/cql_sac.py]: this should be in sac agent

**Comment by Edilmo** [comment]
There are SAC failures all over the place https://bizair.visualstudio.com/Machine%20Teaching/_build/results?buildId=1171521&view=logs&j=b01bcef9-4396-50e3-d06e-2c45d26ab8c8&t=f4fa213f-4fc9-5182-1c60-b401b7d1fc17

I think all of them are related to the lines that I'm recommending to remove in the comments.

File: rllib/tests/agents/parameters.py [hunk 4]
```diff
@@ -371,6 +371,46 @@ def astuple(self):
             threshold=150.0,
             frameworks=[Framework.TensorFlow],
         ),
+        TestAgentParams.for_moab_move_to_center(
```
Comment (Edilmo) [inline rllib/tests/agents/parameters.py:374]: This test is not testing any of the new regularizers. What is the purpose of the test?

File: rllib/agents/cql/cql_sac_tf_policy.py [hunk 5]
```diff
@@ -264,6 +268,10 @@ def cql_loss(policy: Policy, model: ModelV2, dist_class: Type[TFActionDistributi
         # combined in one loss here.
         final_total_loss += alpha_prime_loss
 
+    if vib_regularizer["enable_regularizer"]:
```
Comment (Edilmo) [inline rllib/agents/cql/cql_sac_tf_policy.py]: I think this should be moved to SAC because the VIB Layer is actually managed in SACTFModel.
 If not, you need to move everything to CQLSACTFModel. I prefer the first option, moving to SAC)

**Review by Edilmo — CHANGES_REQUESTED** [review]

File: rllib/agents/cql/cql_sac.py (hunk 3 above)
Comment (abhiksingla) [reply rllib/agents/cql/cql_sac.py]: moved

File: rllib/tests/agents/parameters.py (hunk 4 above)
Comment (abhiksingla) [reply rllib/tests/agents/parameters.py:374]: The regularizer was disabled by mistake. Enabled the test with VIB regularizer.

File: rllib/agents/cql/cql_sac_tf_policy.py (hunk 5 above)
Comment (abhiksingla) [reply rllib/agents/cql/cql_sac_tf_policy.py]: Great suggestion. Moved to SAC

File: rllib/agents/sac/sac_tf_model.py [hunk 6]
```diff
@@ -141,10 +145,13 @@ def build_q_net(self, name, observations, actions):
                 if self.dropout_rate > 0:
                     q_net.append(tf.keras.layers.Dropout(self.dropout_rate))
 
-            q_net += [
-                tf.keras.layers.Dense(
-                    units=self.q_outs, activation=None, name="{}_out".format(name))
-            ]
+            if self.vib_regularizer["enable_regularizer"]:
+                q_net += [VIB_Decoder(output_size=self.q_outs, decoder_layers=self.vib_regularizer["decoder_layers"], dropout_rate=self.dropout_rate)]
+            else:
```
Comment (Edilmo) [inline rllib/agents/sac/sac_tf_model.py]: I know that adding VIB and Dropout sounds weird, but not sure if it's better to allow it anyway. Or at least comment in the config that is not possible to have both and which one has precedence.

**Review by Edilmo — APPROVED** [review]
LGTM. Just a final comment for your consideration.

**Comment by Edilmo** [comment]
The PR still have multiple errors in RLlib. You need to address them before landing.

File: rllib/agents/sac/sac_tf_model.py [hunk 7]
```diff
@@ -118,16 +122,16 @@ def build_action_model(self):
...
                 units=self.action_outs, activation=None, name="action_out")
-        ])
+         ])
         return action_model
 
     def build_q_net(self, name, observations, actions):
         # For continuous actions: Feed obs and actions (concatenated)
         # through the NN. For discrete actions, only obs.
-        if self.dropout_rate>0:
+        if self.dropout_rate > 0 or self.vib_regularizer["enable_regularizer"]:
```
Comment (Edilmo) [inline rllib/agents/sac/sac_tf_model.py]: This if will break all existing tests that have a config without that key. I will check if the config is defined first.

**Review by Edilmo — COMMENTED** [review]
The new IF should be the reason for all the failures.

File: rllib/agents/sac/sac_tf_model.py (hunk 7 above)
Comment (abhiksingla) [reply rllib/agents/sac/sac_tf_model.py]: Thanks. Fixed it

File: rllib/agents/sac/sac_tf_model.py (hunk 6 above)
Comment (abhiksingla) [reply rllib/agents/sac/sac_tf_model.py]: I am not sure if I get this but both dropout and VIB can be enabled simultaneously.

**Review by Edilmo — APPROVED** [review]
Perfect
